src/libraries/Hooks.sol

```diff
@@ -65,42 +83,68 @@ library Hooks {
                 || permissions.beforeDonate != self.hasPermission(BEFORE_DONATE_FLAG)
                 || permissions.afterDonate != self.hasPermission(AFTER_DONATE_FLAG)
+                || permissions.beforeSwapReturnDelta != self.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG)
+                || permissions.afterSwapReturnDelta != self.hasPermission(AFTER_SWAP_RETURNS_DELTA_FLAG)
+                || permissions.afterAddLiquidityReturnDelta != self.hasPermission(AFTER_ADD_LIQUIDITY_RETURNS_DELTA_FLAG)
+                || permissions.afterRemoveLiquidityReturnDelta
+                    != self.hasPermission(AFTER_REMOVE_LIQUIDITY_RETURNS_DELTA_FLAG)
         ) {
             revert HookAddressNotValid(address(self));
         }
     }
 
     /// @notice Ensures that the hook address includes at least one hook flag or dynamic fees, or is the 0 address
     /// @param self The hook to verify
+    /// @return bool True if the hook address is valid
     function isValidHookAddress(IHooks self, uint24 fee) internal pure returns (bool) {
+        // The hook can only have a flag to return a hook delta on an action if it also has the corresponding action flag
+        if (
+            (!self.hasPermission(BEFORE_SWAP_FLAG) && self.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG))
+                || (!self.hasPermission(AFTER_SWAP_FLAG) && self.hasPermission(AFTER_SWAP_RETURNS_DELTA_FLAG))
+                || (
+                    !self.hasPermission(AFTER_ADD_LIQUIDITY_FLAG)
+                        && self.hasPermission(AFTER_ADD_LIQUIDITY_RETURNS_DELTA_FLAG)
+                )
+                || (
+                    !self.hasPermission(AFTER_REMOVE_LIQUIDITY_FLAG)
+                        && self.hasPermission(AFTER_REMOVE_LIQUIDITY_RETURNS_DELTA_FLAG)
+                )
+        ) return false;
         // If there is no hook contract set, then fee cannot be dynamic
         // If a hook contract is set, it must have at least 1 flag set, or have a dynamic fee
         return address(self) == address(0)
             ? !fee.isDynamicFee()
-            : (uint160(address(self)) >= AFTER_DONATE_FLAG || fee.isDynamicFee());
+            : (uint160(address(self)) >= AFTER_REMOVE_LIQUIDITY_RETURNS_DELTA_FLAG || fee.isDynamicFee());
     }
 
-    /// @notice performs a hook call using the given calldata on the given hook
-    /// @return expectedSelector The selector that the hook is expected to return
-    /// @return selector The selector that the hook actually returned
-    function _callHook(IHooks self, bytes memory data) private returns (bytes4 expectedSelector, bytes4 selector) {
+    /// @notice performs a hook call using the given calldata on the given hook that doesnt return a delta
+    /// @return result The complete data returned by the hook
+    function callHook(IHooks self, bytes memory data) internal returns (bytes memory result) {
+        if (msg.sender == address(self)) return "";
```

Comment:
dont love null returns in general, especially since we have to duplicate it in `callHookWithReturnDelta`. feels likely to cause unexpected behavior.. would much prefer not even calling `callHook` if were not actually calling the hook

Reply:
one option that may be better is to return `(bool called, bytes memory result)`

Reply:
oh my bad this was a remnant of an old version while i was waiting for selfCall to be merged. I can just remove this line